structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,290 @@
+				env = append(env, envVar.Key+"="+strings.Replace(envVar.Value, "$"+currentVar, val, -1))
+			} else {
+				t.Errorf("Variable %s not found in image env! Check test config.", currentVar)
+				return nil
+			}
+		} else {
+			env = append(env, envVar.Key+"="+envVar.Value)
+		}
+	}
+	return env
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d *DockerDriver) retrieveFile(t *testing.T, path string) (string, error) {
+	// this contains a placeholder command which does not get run, since
+	// the client doesn't allow creating a container without a command.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Cmd:          []string{"NOOP_COMMAND_DO_NOT_RUN"},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", err
+	}
+
+	tmpFile, err := ioutil.TempFile("", "structure_test")
+	if err != nil {
+		t.Errorf("Error when creating temp file: %s", err.Error())
+		return "", err
+	}
+	stream := bufio.NewWriter(tmpFile)
+
+	err = d.cli.DownloadFromContainer(container.ID, docker.DownloadFromContainerOptions{
+		OutputStream: stream,
+		Path:         path,
+	})
+	if err != nil {
+		t.Errorf("Error when downloading file from container: %s", err.Error())
+		return "", err
+	}
+	stream.Flush()
+	tmpFile.Close()
+	return tmpFile.Name(), nil
+}
+
+func (d *DockerDriver) StatFile(t *testing.T, path string) (os.FileInfo, error) {
+	file, err := d.retrieveFile(t, path)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+
+	f, err := os.Stat(file)
+	if err != nil {
+		return nil, err
+	}
+	if f == nil {
+		return nil, nil
+	}
+	return f, nil
+}
+
+func (d *DockerDriver) ReadFile(t *testing.T, path string) ([]byte, error) {
+	file, err := d.retrieveFile(t, path)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+	return ioutil.ReadFile(file)
+}
+
+func (d *DockerDriver) ReadDir(t *testing.T, path string) ([]os.FileInfo, error) {
+	tmpDir, err := d.retrieveFile(t, path)
+	defer os.RemoveAll(tmpDir)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.ReadDir(tmpDir)
+}
+
+// This method takes a command (in the form of a list of args), and does the following:
+// 1) creates a container, based on the "current latest" image, with the command set as
+// the command to run when the container starts
+// 2) starts the container
+// 3) commits the container with its changes to a new image,
+// and sets that image as the new "current image"
+func (d *DockerDriver) runAndCommit(t *testing.T, env []string, command []string) string {
+	shouldRun := true
+
+	// this is a placeholder command that does not get run, since
+	// the client doesnt allow creating a container without a command.
+	if len(command) == 0 {
```

Comment:
What are the cases where we don't set a command for this function now that you've changed the env var handling? I think you can drop this.

Reply:
ah yeah you're right. removed